### PR TITLE
keystore keymanagement integration

### PR DESCRIPTION
**Pull request by max402**

### CODE REVIEW

File: datasafe-encryption/datasafe-encryption-api/pom.xml [hunk 1]
```diff
@@ -14,12 +14,10 @@
...
             <version>${project.version}</version>
         </dependency>
-		<dependency>
-			<groupId>de.adorsys.keymanagement</groupId>
-			<artifactId>juggler-bouncycastle</artifactId>
-		</dependency>
-
-
+        <dependency>
+            <groupId>de.adorsys.keymanagement</groupId>
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-api/pom.xml]: You should use de.adorsys.keymanagement:api in API
And in impl use de.adorsys.keymanagement:juggler-bouncycastle

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java [hunk 2]
```diff
@@ -6,8 +6,8 @@
 import de.adorsys.datasafe.encrypiton.impl.cmsencryption.decryptors.Decryptor;
 import de.adorsys.datasafe.encrypiton.impl.cmsencryption.decryptors.DecryptorFactory;
 import de.adorsys.datasafe.encrypiton.impl.cmsencryption.exceptions.DecryptionException;
-import de.adorsys.datasafe.encrypiton.impl.keystore.generator.ProviderUtils;
 import de.adorsys.datasafe.types.api.context.annotations.RuntimeDelegate;
+import de.adorsys.keymanagement.bouncycastle.adapter.services.deprecated.generator.ProviderUtils;
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: You should not use classes from `deprecated` package. ProviderUtils should be defined in datasafe

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 3]
```diff
@@ -62,17 +71,21 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
-                .serverKeyPairAliasPrefix(serverKeyPairAliasPrefix)
-                .readKeyPassword(keyStoreAuth.getReadKeyPassword())
-                .secretKeys(secretKeys)
+
+        Juggler juggler = DaggerJuggler.builder().build();
+        EncryptingKeyCreationCfg encConf = keyConfig.getEncrypting();
+        SigningKeyCreationCfg signConf = keyConfig.getSigning();
+        KeySetTemplate template = KeySetTemplate.builder()
+                .generatedEncryptionKeys(Encrypting.with().algo(encConf.getAlgo()).sigAlgo(encConf.getSigAlgo()).keySize(encConf.getSize()).prefix("ENC").password(() -> keyStoreAuth.getReadKeyPassword().getValue()).build().repeat(keyConfig.getEncKeyNumber()))
+                .generatedSigningKeys(Signing.with().algo(signConf.getAlgo()).sigAlgo(signConf.getSigAlgo()).keySize(signConf.getSize()).alias("SIGN-1").password(() -> keyStoreAuth.getReadKeyPassword().getValue()).build().repeat(keyConfig.getSignKeyNumber()))
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: We do not have signing keys in Datasafe

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 4]
```diff
@@ -62,17 +71,21 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
-                .keyCreationConfig(keyConfig)
-                .keyStoreConfig(config)
-                .serverKeyPairAliasPrefix(serverKeyPairAliasPrefix)
-                .readKeyPassword(keyStoreAuth.getReadKeyPassword())
-                .secretKeys(secretKeys)
+
+        Juggler juggler = DaggerJuggler.builder().build();
+        EncryptingKeyCreationCfg encConf = keyConfig.getEncrypting();
+        SigningKeyCreationCfg signConf = keyConfig.getSigning();
+        KeySetTemplate template = KeySetTemplate.builder()
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: Make builder statements one per line:
```
Encrypting.with()
.algo(encConf.getAlgo())
.sigAlgo(encConf.getSigAlgo())
.keySize(encConf.getSize())
.prefix("ENC")
.password(() -> keyStoreAuth.getReadKeyPassword().getValue()).build()
```

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 5]
```diff
@@ -62,17 +71,21 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
+                .generatedEncryptionKeys(Encrypting.with().algo(encConf.getAlgo()).sigAlgo(encConf.getSigAlgo()).keySize(encConf.getSize()).prefix("ENC").password(() -> keyStoreAuth.getReadKeyPassword().getValue()).build().repeat(keyConfig.getEncKeyNumber()))
+                .generatedSigningKeys(Signing.with().algo(signConf.getAlgo()).sigAlgo(signConf.getSigAlgo()).keySize(signConf.getSize()).alias("SIGN-1").password(() -> keyStoreAuth.getReadKeyPassword().getValue()).build().repeat(keyConfig.getSignKeyNumber()))
+                .generatedSecretKeys(secretKeys.keySet().stream().map(it -> Secret.with().prefix(it.getValue()).build()).collect(Collectors.toList()))
                 .build();
+        KeySet keySet = juggler.generateKeys().fromTemplate(template);
 
-        KeyStore userKeyStore = keyStoreGenerator.generate();
+        juggler.withConfig(config);
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: This is dangling statement - it should have created new instance of some class for immutability

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 6]
```diff
@@ -62,17 +71,21 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
+        KeyStore ks = juggler.toKeystore().generate(keySet, () -> keyStoreAuth.getReadKeyPassword().getValue());
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: Extract keyStoreAuth.getReadKeyPassword().getValue() to lambda

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 7]
```diff
@@ -158,21 +171,18 @@ public void removeKey(KeyStoreAccess keyStoreAccess, String alias) {
...
 
     @Override
+    @SneakyThrows
     public byte[] serialize(KeyStore store, String storeId, ReadStorePassword readStorePassword) {
-        return KeyStoreServiceImplBaseFunctions.toByteArray(
-                store,
-                storeId,
-                readStorePassword
-        );
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: Juggler has serialize/deserialize feature

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 8]
```diff
@@ -158,21 +171,18 @@ public void removeKey(KeyStoreAccess keyStoreAccess, String alias) {
...
     @Override
+    @SneakyThrows
     public KeyStore deserialize(byte[] payload, String storeId, ReadStorePassword readStorePassword) {
-        return KeyStoreServiceImplBaseFunctions.loadKeyStore(
-                payload,
-                storeId,
-                config,
-                readStorePassword
-        );
+        KeyStore ks = KeyStore.getInstance(config.getType());
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: Juggler has serialize/deserialize feature

File: datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/KeyPairGeneratorTest.java [hunk 9]
```diff
@@ -10,18 +10,18 @@
 
 class KeyPairGeneratorTest extends WithBouncyCastle {
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/KeyPairGeneratorTest.java]: Probably you wanted to remove this test

File: datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/TestableKeyPairGeneratorImpl.java [hunk 10]
```diff
@@ -1,18 +1,18 @@
...
-
-    public void setWithCA(Boolean ca) {
-        this.withCA = ca;
-    }
-}
+//package de.adorsys.datasafe.encrypiton.impl;
+//
+//import de.adorsys.datasafe.encrypiton.impl.keystore.generator.KeyPairGeneratorImpl;
+//
+//public class TestableKeyPairGeneratorImpl extends KeyPairGeneratorImpl {
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/TestableKeyPairGeneratorImpl.java]: Probably you wanted to remove this test

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java (hunk 2 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/cmsencryption/CMSEncryptionServiceImpl.java]: ok

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 3 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: ok

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 4 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: ok

File: datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/TestableKeyPairGeneratorImpl.java (hunk 10 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/test/java/de/adorsys/datasafe/encrypiton/impl/TestableKeyPairGeneratorImpl.java]: deleted

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 8 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: ok

File: pom.xml [hunk 11]
```diff
@@ -123,6 +123,7 @@
         <spring.framework.boot.version>2.1.5.RELEASE</spring.framework.boot.version>
         <siv-mode.version>1.3.1</siv-mode.version>
         <jackson.version>2.10.0</jackson.version>
+        <keymanagement.version>0.0.1-SNAPSHOT</keymanagement.version>
```
Comment (valb3r) [inline pom.xml]: You should update to 0.0.5

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 12]
```diff
@@ -62,17 +69,28 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
         String serverKeyPairAliasPrefix = UUID.randomUUID().toString();
         log.debug("keystoreid = {}", serverKeyPairAliasPrefix);
-        KeyStoreGenerator keyStoreGenerator = KeyStoreGenerator.builder()
-                .keyCreationConfig(keyConfig)
-                .keyStoreConfig(config)
-                .serverKeyPairAliasPrefix(serverKeyPairAliasPrefix)
-                .readKeyPassword(keyStoreAuth.getReadKeyPassword())
-                .secretKeys(secretKeys)
+
+        BCJuggler juggler = DaggerBCJuggler.builder().build();
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: You should provide Juggler through EncryptionModule. Also its builder accepts EncryptionConfig, so you can use 
it too.
```
@Provides
Juggler bcJuggler(Config config) {
return DaggerBCJuggler.builder().withConfig(config).build();
}

File: pom.xml (hunk 11 above)
Comment (max402) [reply pom.xml]: ok

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 12 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: ок

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 13]
```diff
@@ -62,39 +70,39 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
+                        .build()
+                        .repeat(keyConfig.getEncKeyNumber()))
+                .generatedSecretKeys(secretKeys.keySet().stream()
+                        .map(it -> Secret.with().prefix(it.getValue()).build())
+                        .collect(Collectors.toList()))
                 .build();
+        KeySet keySet = juggler.generateKeys().fromTemplate(template);
 
-        KeyStore userKeyStore = keyStoreGenerator.generate();
+        KeyStore ks = juggler.toKeystore().withConfig(config).generate(keySet, passSupplier);
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: ```.withConfig(config)```
Is unnecessary - we already have default in Juggler
See 
https://github.com/adorsys/datasafe/pull/181/files#diff-88b877cc34976bb2cfb6d55188a7768eR39

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 14]
```diff
@@ -157,21 +165,14 @@ public void removeKey(KeyStoreAccess keyStoreAccess, String alias) {
...
-    public KeyStore deserialize(byte[] payload, String storeId, ReadStorePassword readStorePassword) {
-        return KeyStoreServiceImplBaseFunctions.loadKeyStore(
-                payload,
-                storeId,
-                config,
-                readStorePassword
-        );
+    @SneakyThrows
+    public KeyStore deserialize(byte[] payload, ReadStorePassword readStorePassword) {
+        return juggler.serializeDeserialize().withConfig(config).deserialize(payload, readStorePassword::getValue);
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: Unneeded withConfig
https://github.com/adorsys/datasafe/pull/181/files#diff-88b877cc34976bb2cfb6d55188a7768eR39

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 13 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: removed

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java (hunk 14 above)
Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: removed

File: datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java [hunk 15]
```diff
@@ -62,39 +70,39 @@ public KeyStore createKeyStore(KeyStoreAuth keyStoreAuth,
...
+                        .build()
+                        .repeat(keyConfig.getEncKeyNumber()))
+                .generatedSecretKeys(secretKeys.keySet().stream()
+                        .map(it -> Secret.with().prefix(it.getValue()).build())
+                        .collect(Collectors.toList()))
                 .build();
+        KeySet keySet = juggler.generateKeys().fromTemplate(template);
 
-        KeyStore userKeyStore = keyStoreGenerator.generate();
+        KeyStore ks = juggler.toKeystore().generate(keySet, passSupplier);
```
Comment (valb3r) [inline datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: passSupplier is not needed here - you pass it in KeySetTemplate

Comment (max402) [reply datasafe-encryption/datasafe-encryption-impl/src/main/java/de/adorsys/datasafe/encrypiton/impl/keystore/KeyStoreServiceImpl.java]: done

**Pull request merged**